sigmf/archive.py

```diff
@@ -0,0 +1,140 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Create and extract SigMF archives."""
+
+import os
+import shutil
+import tarfile
+import tempfile
+
+from . import error
+
+
+SIGMF_ARCHIVE_EXT = ".sigmf"
+SIGMF_METADATA_EXT = ".sigmf-meta"
+SIGMF_DATASET_EXT = ".sigmf-data"
+
+
+class SigMFArchive(object):
+    """Archive a SigMFFile.
+
+    A `.sigmf` file must include both valid metadata and data. If metadata
+    is not valid, raise `SigMFValidationError`. If `self.data_file` is not
+    set, raise `SigMFFileError`.
+
+    Parameters:
+
+      sigmffile -- A SigMFFile object with valid metadata and data_file
+
+      name      -- path to archive file to create. If file exists, overwrite.
+                   If `name` doesn't end in .sigmf, it will be appended.
+                   For example: if `name` == "/tmp/archive1", then the
+                   following archive will be created:
+                       /tmp/archive1.sigmf
+                       - archive1/
+                         - archive1.sigmf-meta
+                         - archive1.sigmf-data
+
+      fileobj   -- If `fileobj` is specified, it is used as an alternative to
+                   a file object opened in binary mode for `name`. It is
+                   supposed to be at position 0. `name` is not required, but
+                   if specified will be used to determine the directory and
+                   file names within the archive. `fileobj` won't be closed.
+                   For example: if `name` == "archive1" and fileobj is given,
+                   a tar archive will be written to fileobj with the
+                   following structure:
+                       - archive1/
+                         - archive1.sigmf-meta
+                         - archive1.sigmf-data
+
+    """
+    def __init__(self, sigmffile, name=None, fileobj=None):
+        self.sigmffile = sigmffile
+        self.name = name
+        self.fileobj = fileobj
+
+        self._validate()
+
+        archive_name = self._get_name()
+        sigmf_fd = self._open()
+        sigmf_archive = tarfile.TarFile(mode="w",
+                                        fileobj=sigmf_fd)
+        tmpdir = tempfile.mkdtemp()
+        sigmf_md_filename = archive_name + SIGMF_METADATA_EXT
+        sigmf_md_path = os.path.join(tmpdir, sigmf_md_filename)
+        sigmf_data_filename = archive_name + SIGMF_DATASET_EXT
+        sigmf_data_path = os.path.join(tmpdir, sigmf_data_filename)
+
+        with open(sigmf_md_path, "w") as mdfile:
```

Comment:
In this case, `sigmf_md_path` is being created under `tmpdir`, which is guaranteed to be "readable, writable, and searchable only by the creating user ID." ([mkdtemp](https://docs.python.org/3.5/library/tempfile.html#tempfile.mkdtemp)), so I _think_ we're 100% safe here.